### PR TITLE
* Add app story progress to logged in account when redeem or bind

### PR DESCRIPTION
**Pull request by mileswong**

### CODE REVIEW

File: modmod/operations/user.py [hunk 1]
```diff
@@ -74,3 +75,26 @@ def handle_membership_update(user, original_transaction_id, expire_timestamp, \
         user.expire_date = new_expire_date
     return 'ok'
 
+def handle_anonymous_user_app_story_progress(is_existing_user, prev_user, new_user):
```
Comment (nwingt) [inline modmod/operations/user.py]: prev_user => prev_user_email

File: modmod/operations/user.py [hunk 2]
```diff
@@ -74,3 +75,26 @@ def handle_membership_update(user, original_transaction_id, expire_timestamp, \
...
+
+    if old_user.is_anonymous:
+        old_user_progress = UserReadOiceProgressQuery(DBSession).fetch_by_user_id(old_user.id)
+        if is_existing_user:
+            new_user_progress = UserReadOiceProgressQuery(DBSession).fetch_by_user_id(new_user.id)
+
+            new_user_read_oice_ids = set([p.oice_id for p in new_user_progress])
+            for progress in old_user_progress:
+                if progress.oice_id in new_user_read_oice_ids:
+                    updated_progress = next((p for p in new_user_progress if p.oice_id == progress.oice_id))
```
Comment (nwingt) [inline modmod/operations/user.py]: maybe turn `new_user_progress` to dict where key is `oice_id` and `is_finished` is value for checking

File: modmod/operations/user.py [hunk 3]
```diff
@@ -74,3 +75,26 @@ def handle_membership_update(user, original_transaction_id, expire_timestamp, \
...
 
+def handle_anonymous_user_app_story_progress(is_existing_user, prev_user, new_user):
+    # add story progress of anonymous user to the account binded / redeemed
+    old_user = UserQuery(DBSession).fetch_user_by_email(email=prev_user).one_or_none()
+    if not old_user:
+        raise HTTPForbidden
+
+    if old_user.is_anonymous:
+        old_user_progress = UserReadOiceProgressQuery(DBSession).fetch_by_user_id(old_user.id)
+        if is_existing_user:
```
Comment (nwingt) [inline modmod/operations/user.py]: Can `is_existing_user` be replaced by `old_user.id != new_user.id`?

Comment (mileswong) [reply modmod/operations/user.py]: no, is_existing_user indicates whether the new_user is logged in before

**Pull request merged**